### ISSUE TITLE
Integrate Red Hat stack + e2es into jenkins run 

### ISSUE THREAD

**Issue by timothysc** [kubernetes/kubernetes#10934]
For some time now we have maintained a bare-metal RHEL-Atomic cluster on our end which served two primary purposes. 
1. Evaluate performance & HA, and feedback issues upstream.
2. Verify stack sanity by running e2e gauntlet on our stack.  

#2 has proven to be untenable, as we seem to eat a large number of changes that break for various reasons, and we're left hunting to feedback what has broken the system. 

Going forwards, we would like to add the automated startup and provisioning of our stack to run the e2e-tests and to red-flag commits before it reaches us downstream. 

The goal of this issue is to rationalize how we intend to solve this and integrate with upstream.  

/cc @eparis @jayunit100 @rrati @pmorie 
/cc @quinton-hoole @roberthbailey @ixdy

**Comment by zmerlynn** [comment]
:+1:

**Comment by jayunit100** [comment]
We have a cloud friendly vagrant-ansible recipe which can be extended to provision GCE boxes #10706 .  We just need to 

- add GCE support to vagrantfile
- add E2E playbook to contrib/ansible/

Then from their google guys can take over and deploy it on jenkins?

**Comment by timothysc** [comment]
@jayunit100 Well it's more then that really... We need to maintain the images and be able to rev the components of the stack as well.  

For example, I would ideally like to have RHEL 7 images running 
- commit of k8s
- etcd - latest 
- flannel - latest 
- docker - latest 

Ideally what I would like is an upstream atomic image that kept up to date on the latest, then just update the image itself and not worry about managing the versions. 

cc - @cgwalters, @wattsteve

**Comment by jayunit100** [comment]
agree +1

**Comment by pmorie** [comment]
@timothysc We need to have coverage of SELinux enforcing / disabled.  Optional: permissive.

**Comment by pmorie** [comment]
Also touches:

1.  https://github.com/GoogleCloudPlatform/kubernetes/issues/6605
2.  https://github.com/GoogleCloudPlatform/kubernetes/issues/4871

**Comment by timothysc** [comment]
@pmorie why would we ever care about !enforcing?

**Comment by ncdc** [comment]
@timothysc some users run with SELinux disabled, and it is possible to break things if you assume it's always at least permissive (e.g. try to call `chcon`).

**Comment by timothysc** [comment]
@ixdy - per our discussion last week do you have an outline of your current configuration and how we could potentially add-on?

**Comment by pmorie** [comment]
@timothysc Actually, we have coverage of disabled from the existing suite, so I don't think we do care about explicitly testing with selinux disabled on our stack.  However, in the interest of full disclosure, I was originally concerned about the bugs that you will hit with SELinux disabled that aren't problems when it's enabled (enforcing or permissive).  For example, libcontainer selinux API isn't built for !linux at all.  If you link to it without special casing the build platform, you'll get errors.

**Comment by pmorie** [comment]
@ixdy @timothysc @eparis 

How actually do we want the integration with github to work?

So far, we have discussed the following for the github flow on the RH stack:

1.  Build
2.  Unit tests
3.  Integration tests
4.  Deploy with vagrant+ansible
5.  Subset of E2E tests

Do we want another bot to require a project owner to authorize the tests/whitelist people, or do we want something closer to travisci/shippable where the tests run automatically?

**Comment by quinton-hoole** [comment]
cc @davidopp

**Comment by cgwalters** [comment]
I'm actively working on a CentOS Atomic SIG tree build which will pull more directly from upstream git, including k8s.  Will update here as that progresses.

**Comment by davidopp** [comment]
@jdef @karlkfi you guys will probably be interested in this discussion

**Comment by childsb** [comment]
Can we run the full set of E2E on some sort of interval?  Maybe every Friday/over the weekend.

**Comment by pmorie** [comment]
@childsb The current plan is to run the full suite on a nightly basis.

**Comment by ixdy** [comment]
Sorry for the delay in writing this up.                                          
                                                                                 
Right now we run a Jenkins instance that runs e2e tests both before and after PRs are merged.
                                                                                 
The Jenkins projects run the scripts in hack/jenkins, either setting some environment variables before calling the script, or letting the script auto-configure based on the job name. 
                                                                                 
The per-PR workflow is a little bit more straightforward, so I'll describe it first.
                                                                                 
We use the [GitHub pull request Jenkins plugin](https://wiki.jenkins-ci.org/display/JENKINS/GitHub+pull+request+builder+plugin) to find PRs to test. This is hooked up to the @k8s-bot user, which allows it to comment on PRs.
                                                                                 
Because we didn't want to publicly expose our Jenkins instance, it polls GitHub every 2 minutes instead of using webhooks. Whenever there is a new PR or new commits on a PR, it builds and runs the e2e tests, so long as the PR author is in the whitelist (currently defined to be the list of users with commit access).

Running tests consists of setting `KUBE_SKIP_PUSH_GCS=y` and `KUBE_RUN_FROM_OUTPUT=y `to prevent any build artifacts from being uploaded to Google Cloud Storage, and then running
```                                                                              
./hack/jenkins/build.sh                                                          
./hack/jenkins/e2e.sh                                                            
make clean                                                                       
```                                                                              
                                                                                 
This job is named `kubernetes-pull-build-test-e2e-gce`, and so also sets a number of variables as defined in `hack/jenkins/e2e.sh`. We don't run all of the e2e tests - just those that are deemed stable to run in Ginkgo's parallel mode. (This list is also defined in `e2e.sh`.)

After the test completes, we upload the build log and everything under `_artifacts/` to GCS so that non-Googler can see the results.                      
                                                                                 
It'd probably be easy enough for anyone else to set up a separate Jenkins instance which does much the same as ours does, though there are a few things to note: 
- Each invocation takes around 18-20m to run. We have the PR builder set up to run up to 4 builds in parallel, which so far seems around the right level to keep up. When we first started up the PR builder, we basically had to drop all open PRs, since it was likely going to take us days to get through the backlog.
- When I first set up the PR builder plugin, it seemed to require having commit access to the repo in order to update the commit statuses. I'm not sure why this was necessary - as Travis/Shippable don't need this - but maybe it's related to having webhooks for those set up?
- It's definitely necessary to use authenticated requests to GitHub's API; the unauthenticated limit is 60 qph, and we easily exceed that. (This limit helped me discover a bug in the PR builder plugin at one point.)
- Testing everything against a separate repo first is probably a good idea - I did this extensively in k8s-bot/kubernetes#1 and k8s-bot/kubernetes#2, for example.                                                                                                                                                   
- Probably other things I'm forgetting.                                          
                                                                                 
I should probably upload our full Jenkins configuration somewhere, since it's not possible to pull everything into `e2e.sh`, and because describing it all in text loses detail, anyway.

Long term we do need to invest in some sort of aggregated dashboard, because relying on PR comments or commit status along is probably going to be insufficient.

Post-PR workflow coming up in a later comment.

**Comment by pmorie** [comment]
Update: I have a test jenkins instance set up, a new bot github account, and a job set up that monitors my fork and runs a stub build for each update to a PR set up.  Next step is replacing the stub build with the real build steps.

@ixdy We should start thinking about how we want to ultimately integrate this into the github flow.  I would really like to be able to drive the different bots from the same comments (ie, don't require mentioning **both** k8s-bot and rh-kube-bot), and use the same whitelist.

@ixdy created https://github.com/jenkinsci/ghprb-plugin/issues/160 for dirving the whitelist from a GH org.  I'm sort of wondering if we could make a team for github bots for the kubernetes org and use that to drive the 'ok to test' messages.

**Comment by timothysc** [comment]
+1 to have an automation 
[ok to test]

**Comment by pmorie** [comment]
@quinton-hoole @ixdy We've started talking about where to store test results, and I wondered if we could use the GCS storage already being used for Google's internal jenkins results for the RH jenkins results as well.  Would that be possible?

**Comment by quinton-hoole** [comment]
@pmorie I think that it would be preferable to have a separate GCS bucket for your test results (hence separate permissions, quota, billing etc). I'll chat to you offline on how best to achieve that.

**Comment by pmorie** [comment]
@quinton-hoole Cool, thanks.

**Comment by timothysc** [comment]
@ingvagabund is this done?

**Comment by ingvagabund** [comment]
@timothysc not yet, integrating storage tests into the Job

**Comment by timothysc** [comment]
re-opening per: @ingvagabund 's comments.

**Comment by fejta** [comment]
Thanks for reopening!

**Comment by ingvagabund** [comment]
So here we are. I have just launched the CI on github.com/kubernetes/kubernetes [1].
In short what is currently available:

- we use redhat-k8s-bot bot to trigger the CI, the CI is limited to PRs with "team/node" label
- all results/artifacts are uploaded to gs bucket
- we can run up to 4 CIs in parallel
- running unit tests, integration tests and test-cmd
- cluster deployed with ansible, cluster made of RHEL images
- running Conformance and Volumes e2e tests (concentrating on node components)
- the CI has to be triggered manually at this point, later it will get triggered automatically
- currently these people can trigger the testing, tag them in if you have a PR you want to try it on: ingvagabund, sttts, ncdc, pmorie, eparis, derekwaynecarr. I am working on extending the list with everyone in kubernetes organization (GHPRB has allow-whitelist-orgs-as-admins parameter which I have discovered just now).

TL;DR here [2].

[1] https://github.com/kubernetes/kubernetes/pull/26551
[2] https://groups.google.com/forum/#!topic/kubernetes-dev/wA_Vfm3yCEY

**Comment by ingvagabund** [comment]
@timothysc I remember you had other tests in mind we could run in the CI. Can you recall?

**Comment by ixdy** [comment]
cc @kubernetes/sig-testing

**Comment by timothysc** [comment]
@ingvagabund where are we at here?

**Comment by timothysc** [comment]
RH has their own testing via OS, closing this issue.
